Update singleton doc (#443)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/singleton-actors.md

# Singleton Actors

> Singleton Actors allow a single source of truth for both operations and data across a multi-server simulation. They are server-side authoritative [Actors](https://docs.unrealengine.com/en-us/Programming/UnrealArchitecture/Actors) that are restricted to one instantiation on SpatialOS.

Reviewer: I think we can just say "Unreal Actors" here to make it clear it's not any funny business.
So: 
"They are server-side authoritative Unreal Actors that..."

> Singleton Actors allow a single source of truth for both operations and data across a multi-server simulation. They are server-side authoritative [Unreal Actors](https://docs.unrealengine.com/en-us/Programming/UnrealArchitecture/Actors) that are restricted to one instantiation on SpatialOS.

Reviewer 2: Let's give a more layman example justifying why we need this. "For example, if you are implementing a scoreboard, there needs to be only one of them in your world; ensuring this behavior in our multi-server paradigm requires a few additional steps".

> Singleton Actors allow a single source of truth for both operations and data across a multi-server simulation. They are server-side authoritative [Unreal Actors](https://docs.unrealengine.com/en-us/Programming/UnrealArchitecture/Actors) that are restricted to one instantiation on SpatialOS.

Author: ```suggestion
Singleton Actors allow a single source of truth for both operations and data across a multi-server simulation. They are server-side authoritative [Unreal Actors](https://docs.unrealengine.com/en-us/Programming/UnrealArchitecture/Actors) that are restricted to one instantiation on SpatialOS. For example, if you are implementing a scoreboard, you'd most likely only want there to be one of them in your world. Ensuring this behavior in a multi-server paradigm requires a few additional steps which can be easily facilitated through the Singleton Actor.
```

There are two kinds of Singleton Actors:

* **Public Singleton Actors** - Singleton Actors which are replicated to server-workers and client-workers. `GameState` is a Public Singleton Actor.

> * **Private Singleton Actors** - Singleton Actors which are replicated to server-workers, but not accessible to client-workers. `GameMode` is a Private Singleton Actor.

Reviewer: Nice.

Reviewer: Link to glossary term here (I have put this in backticks so you can see the link but we don't need the backticks in the text). I think in this instance, we should put the links in for both entries - even though they are the same link. (Normally, you only do it once but it feels more right to do it each time here??!?)
`....which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers) .`

AND

`replicated to [server-workers]({{urlRoot}/content/glossary#workers), but not accessible to [client-workers]({{urlRoot}/content/glossary#workers).

> * **Public Singleton Actors** - Singleton Actors which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers). `GameState` is a Public Singleton Actor.

Reviewer 3: maybe link to the GameState API page in unreal or of we have a more relevant page

Reviewer 3: Would apply to the other places where we refer to unreal classes too

> * **Public Singleton Actors** - Singleton Actors which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers). `GameState` is a Public Singleton Actor.

Reviewer 2: "For example, `AGameState` is a Public Singleton Actor"

>  * **Public Singleton Actors** - Singleton Actors which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers). `GameState` is a Public Singleton Actor.

Author: ```suggestion
* **Public Singleton Actors** - Singleton Actors which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers). `GameState` is a Public Singleton Actor.
```

> * **Private Singleton Actors** - Singleton Actors which are replicated to server-workers, but not accessible to client-workers. `GameMode` is a Private Singleton Actor.

Author: ```suggestion
* **Private Singleton Actors** - Singleton Actors which are replicated to [server-workers]({{urlRoot}/content/glossary#workers), but not accessible to [client-workers]({{urlRoot}/content/glossary#workers). `GameMode` is a Private Singleton Actor.
```

> * **Private Singleton Actors** - Singleton Actors which are replicated to [server-workers]({{urlRoot}/content/glossary#workers), but not accessible to [client-workers]({{urlRoot}/content/glossary#workers). `GameMode` is a Private Singleton Actor.

Reviewer 2: "For example, `AGameMode` (link to UE4) is a Private Singleton Actor"

> * **Public Singleton Actors** - Singleton Actors which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers). `GameState` is a Public Singleton Actor.

Author: ```suggestion
* **Public Singleton Actors** - Singleton Actors which are replicated to [server-workers and client-workers]({{urlRoot}/content/glossary#workers). [AGameState](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode) is a Public Singleton Actor.
```

You can define any class as a Singleton Actor.

Each server-worker should instantiate their own local version of each Singleton Actor. For `GameMode` and `GameState`, Unreal Engine does this automatically. Client-workers receive public singletons from the server-workers via the normal Actor replication lifecycle.

> 

Reviewer: Singletons . <- I think we made them proper nouns.

What do we mean? "Singleton objects"? (As they are an instatiation of a Singleton class?) Or should we be saying "Singleton Actor"?)

> * **Private Singleton Actors** - Singleton Actors which are replicated to [server-workers]({{urlRoot}/content/glossary#workers), but not accessible to [client-workers]({{urlRoot}/content/glossary#workers). `GameMode` is a Private Singleton Actor.

Author: ```suggestion
* **Private Singleton Actors** - Singleton Actors which are replicated to [server-workers]({{urlRoot}/content/glossary#workers), but not accessible to [client-workers]({{urlRoot}/content/glossary#workers). [AGameMode](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode) is a Private Singleton Actor.
```

> Each server-worker should instantiate their own local version of each Singleton Actor. For `GameMode` and `GameState`, Unreal Engine does this automatically. Client-workers receive public singletons from the server-workers via the normal Actor replication lifecycle.

Author: ```suggestion
Each server-worker should instantiate their own local version of each Singleton Actor. For `GameMode` and `GameState`, Unreal Engine does this automatically. Client-workers receive Public Singletons Actors from the server-workers via the normal Actor replication lifecycle.
```

> Each server-worker should instantiate their own local version of each Singleton Actor. For `GameMode` and `GameState`, Unreal Engine does this automatically. Client-workers receive Public Singletons Actors from the server-workers via the normal Actor replication lifecycle.

Reviewer 3: AGameMode && AGameState?

> Each server-worker should instantiate their own local version of each Singleton Actor. For `GameMode` and `GameState`, Unreal Engine does this automatically. Client-workers receive Public Singletons Actors from the server-workers via the normal Actor replication lifecycle.

Author: ```suggestion
Each server-worker should instantiate their own local version of each Singleton Actor. For `AGameMode` and `AGameState`, Unreal Engine does this automatically. Client-workers receive Public Singletons Actors from the server-workers via the normal Actor replication lifecycle.
```

> Due to Unreal server-workers spawning their own instances of each Singleton Actor, proper replication and authority management of Singleton Actors becomes a bit tricky. To solve this issue, we have introduced the concept of a Global State Manager (GSM) to enable proper replication of Singleton Actors. The GSM solves the problem of replicating Singleton Actors by only allowing the server-worker with [authority](https://docs.improbable.io/reference/latest/shared/glossary#read-and-write-access-authority) over the GSM to execute the initial replication of these Actors. All other server-workers will then link their local Singleton Actors to their respective SpatialOS entity. Because of this, you must update your snapshot whenever adding a new Singleton Actor to your project.

Reviewer: In the context of the other new docs, this is now a bit confusing...
"Due to Unreal server-workers spawning their own instances of each Singleton Actor,". 

Do we mean native Unreal networking servers or do we mean SpatialOS server-workers?

---
Glossary link for authority:
[authority]({{urlroot}}/content/glossary#authority)

Author: Link to our glossary authority definition, instead of the improbable one?

Author: re: your first comment, both? In native unreal, the single server-worker, would create all the singletons. In Spatial Unreal, each server-worker creates the singletons, so we need additional logic to link them together, and ensure that only one of the server-workers is actually authoritative over them.

Reviewer: RE: Glossary - 
Yes, keep peeps in the "padded cell" of Unreal land. The Glossary links out to the SpatialOS docs - and has a note that you are entering SpatialOS land where things might be a bit different as the content assumes you are using the SpatialOS (worker) SDK.

> Due to Unreal server-workers spawning their own instances of each Singleton Actor, proper replication and authority management of Singleton Actors becomes a bit tricky. To solve this issue, we have introduced the concept of a Global State Manager (GSM) to enable proper replication of Singleton Actors. The GSM solves the problem of replicating Singleton Actors by only allowing the server-worker with [authority](https://docs.improbable.io/reference/latest/shared/glossary#read-and-write-access-authority) over the GSM to execute the initial replication of these Actors. All other server-workers will then link their local Singleton Actors to their respective SpatialOS entity. Because of this, you must update your snapshot whenever adding a new Singleton Actor to your project.

Author: ```suggestion
Due to server-workers spawning their own instances of each Singleton Actor, proper replication and authority management of Singleton Actors becomes a bit tricky. To solve this issue, we have introduced the concept of a Global State Manager (GSM) to enable proper replication of Singleton Actors. The GSM solves the problem of replicating Singleton Actors by only allowing the server-worker with [authority]({{urlroot}}/content/glossary#authority) over the GSM to execute the initial replication of these Actors. All other server-workers will then link their local Singleton Actors to their respective SpatialOS entity. Because of this, you must update your snapshot whenever adding a new Singleton Actor to your project.
```

}
```

The code snippet below shows how to tag a native C++ class with the appropriate Private Singleton identifiers.

```
UCLASS(SpatialType=(Singleton, ServerOnly))
class TESTSUITE_API AExampleGameGameMode : public AGameModeBase
{
  GENERATED_BODY()
  ...
}
```

> To tag a blueprint class as a Public Singleton, open the class in the blueprint editor and navigate to the `Class Settings`. In the `Advanced` section inside `Class Options`, tick the `Spatial Type` checkbox and add `Singleton` to the `Spatial Description` textbox. To tag a blueprint class as a Private Singleton, follow the same steps, but add `ServerOnly` to the `Spatial Description` textbox.

Reviewer: Blueprint  <-it's a proper noun according to Unreal.
Editor <- the Blueprint Editor is also a proper noun, according to Unreal.

Reviewer: Also "check" rather than "tick" because we are American now. 
Suggest edits thus:

"To tag a Blueprint class as a Public Singleton, open the class in the Blueprint Editor and navigate to the `Class Settings`. In the `Advanced` section inside `Class Options`, check the `Spatial Type` checkbox and add `Singleton` to the `Spatial Description` textbox. To tag a Blueprint class as a Private Singleton, follow the same steps, but add `ServerOnly` to the `Spatial Description` textbox.

> To tag a blueprint class as a Public Singleton, open the class in the blueprint editor and navigate to the `Class Settings`. In the `Advanced` section inside `Class Options`, tick the `Spatial Type` checkbox and add `Singleton` to the `Spatial Description` textbox. To tag a blueprint class as a Private Singleton, follow the same steps, but add `ServerOnly` to the `Spatial Description` textbox.

Author: ```suggestion
To tag a Blueprint class as a Public Singleton, open the class in the Blueprint Editor and navigate to the `Class Settings`. In the `Advanced` section inside `Class Options`, check the `Spatial Type` checkbox and add `Singleton` to the `Spatial Description` textbox. To tag a Blueprint class as a Private Singleton, follow the same steps and add `ServerOnly` to the `Spatial Description` textbox.
```

> This is an example of what your blueprint `Class Options` should look like if you've tagged it as a Private Singleton:

Reviewer: Blueprint

> ![Singleton blueprint]({{assetRoot}}assets/screen-grabs/blueprint_singleton.png)

Reviewer 3: Doesn't this png have to be in the PR too?

> And that's it! You have successfully specified a Singleton Actor. Make sure you generate schema and create a new snapshot using the [SpatialOS GDK for Unreal toolbar]({{urlRoot}}/content/toolbar).

Reviewer: Suggest link to the glossary and using the slightly shorter name for the toolbar (have put in backticks to keep the link exposed. I am not suggesting this is a good look for this bit of text). :

`Make sure you generate [schema]({{urlRoot}}/content/glossary#schema) and create a new [snapshot]( ({{urlRoot}}/content/generating-a-snapshot) using the [SpatialOS GDK toolbar]({{urlRoot}}/content/toolbar).`
